docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,298 @@
+
+This RFC proposes changing the existing `lift` syntax as a way to capture the necessary type information during preflight.
+An important feature of this syntax is that it's entirely statically analyzable, so anything we need to do with it can be done before-preflight.
+The existing `lift` syntax is currently intended to be used only inside inflights to explicitly lift preflight data into the defined scope:
+
+```wing
+let b1 = new cloud.Bucket();
+inflight () => {
+  let var b = b1;
+  lift { b: [put] } {
+    b.put("key", "value");
+  }
+}
+```
+
+With some changes, with syntax could also be used as a way to create a preflight container for lifted data.
+
+```wing
+let b = new cloud.Bucket();
+let lifts = lift { b: [put] };
+//  ^^^^^ is a std.LiftQualificationsContainer
+```
+
+This is very similar to the `lift` mechanism in `@wingcloud/framework`
+
+```ts
+const b = new cloud.Bucket(app, "Bucket");
+const lifts = lift({ bucket }).grant({ bucket: ["put"] });
+```
+
+In the [RFC for explicit lift qualifications](https://github.com/winglang/wing/blob/main/docs/contributing/999-rfcs/2024-03-14-explicit-lift-qualification.md), there is a mention of a `std.LiftQualificationsContainer`.
+
+If this data structure is augmented to also include type information, it could be used to generate the client during preflight.
+
+```wing
+let b = new cloud.Bucket();
+client.add(lift { b: [put] });
+```
+
+An initial implementation of this can re-use the existing `.d.ts` generation, and the std.LiftQualificationsContainer can store a path to that generated file to be used as-is when generating the full client in preflight.
+
+## Example Usage
+
+### Kubernetes
+
+When generating a client for use in a Kubernetes pod, ConfigMaps are a standard mechanism for storing configuration data.
+To resolve the context, the ConfigMap would be mounted to the pod and the client would read the data from the mounted file:
+
+```wing
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return fs.readJson("/var/run/client-ctx.json");
+  },
+);
+```
+
+If the kubernetes manifests for the cluster is managed in wing, the ConfigMap can be directly created to store the context data:
+
+```wing
+new k8s.ConfigMap(
+  name: "client-ctx",
+  data: client.context,
+);
+```
+
+Otherwise, the data can be mapped similarly to the other methods like SSM or buckets.
+If you're using terraform/cfn, an output can be created with the context be managed manually after deployment.
+
+```wing
+// cdktf
+new cdktf.Output(
+  name: "ClientContext",
+  value: client.bindContext
+);
+
+/// aws-cdk-lib
+new awscdk.CFNOutput(
+  name: "ClientContext",
+  value: client.bindContext
+);
+```
+
+This could be used to create a user/role that is eventually usable within the k8s cluster as a service account.
+
+### AWS Lambda
```

Comment:
In the sentence here I do mention creating a role and using a k8s service account. I didn't write it out because at that point it feels out of scope for the RFC.
I can specifically add a note about `aws.generatePolicyStatements(client.lifts)` as you mentioned, but writing out the actual usage of policy statements wouldn't really change how the clients works.